### DISCUSSION TITLE
Using UtteranceEnd and Endpointing

### DISCUSSION THREAD

**Discussion by celestk** [deepgram/community#980] [General help] [answered]
Hi Team,

I'm building a conversational chatbot and using speech_final = true to detect end of speech. It works well in most cases. My configs:
`    interim_results: true,
    smart_format: true,
    endpointing: 800,
    utterance_end_ms: 2000,
    filler_words: true`

However there are some cases whereby the speech_final doesn't come through, probably due to background noise. 

I understand that according the docs (https://developers.deepgram.com/docs/understanding-end-of-speech-detection#using-utteranceend-and-endpointing), we can "trigger if you receive an UtteranceEnd message with no preceding speech_final=true message and send the last-received transcript for further processing".

My question is when do i start tracking this particular Utterance End event? 

I notice when speech_final is working fine, i will get an Utterance End event in the next 2nd or 3rd or 4th transcript event, which i should be ignoring. Am i right to say that i need to start tracking the next Utterance End event?

Thank you.

**Comment by deepgram-community** [comment]
Hey there! It looks like you haven't connected your GitHub account to your Deepgram account. You can do this at https://community.deepgram.com - being verified through this process will allow our team to help you in a much more streamlined fashion.

**Comment by deepgram-community** [comment]
It looks like we're missing some important information to help debug your issue. Would you mind providing us with the following details in a reply?

- The programming language you are working in (e.g. JavaScript, Python).
- A request ID that triggered your error or issue.

**Comment by jkroll-deepgram** [comment, marked as answer]
Hi @celestk, as you mention, an `UtteranceEnd` event will come after a `speech_final` event, if the `speech_final` event does occur. If it doesn't, you'll make use of the next `UtteranceEnd` event.

In cases when speech_final fires: 
- {transcript with `speech_final=false`}
- {transcript with `speech_final=true`} (informs you that end of speech has been detected!)
- {`UtteranceEnd`} (ignored)

In cases when speech_final does not fire:
- {transcript with `speech_final=false`}
- {transcript with `speech_final=false`} (end of speech is missed due to background noise)
- {`UtteranceEnd`} (informs you that end of speech has been detected!)

**Reply by deva-gopalani** [reply]
@jkroll-deepgram is it possible to first receive an UtteranceEnd event and then receive a speech_final event for the same transcript?

**Reply by jkroll-deepgram** [reply]
Hi @deva-gopalani, typically no. Speech_final uses endpointing, which is at the level of tens to hundreds of milliseconds. Utterance End is greater than 1 second. So endpointing/speech_final would fire _first_, and UtteranceEnd would follow.
